File: chapter1/python/oneAway.py
def oneAway(str_1, str_2):

    # 1. Determine the longer string
    len_1 = len(str_1)
    len_2 = len(str_2)
    if len_1 > len_2:
        big = str_1
        small = str_2
    else:
        big = str_2
        small = str_1

    # 2. loop on the longer string
    fp = 0
    while fp < len(big) - 1:
        # if these two letters match, continue
        if big[fp] == small[fp]:
            fp += 1
            continue
        else:
            # if the difference is in the first or the last, remove it from both strings
            if fp == 0 or fp == len(big) - 1:
                new_str = big[:fp] + big[fp+1:]
                small = small[:fp] + small[fp+1:]
            else:
                # if the difference is in the middle, just remove if from the bigger string
                new_str = big[:fp] + big[fp+1:]
            # now return whether they are equal or not
            return new_str == small
    # you will only reach this if there is no difference found, so return True
    return True


# TESTS
print(oneAway("pale", "ple"))  # True
print(oneAway("pales", "pale"))  # True
print(oneAway("pale", "bale"))  # True
print(oneAway("pale", "bake"))  # False
print(oneAway("balb", "balp"))  # True
print(oneAway("balp", "balp"))  # True
print(oneAway("", ""))  # True


# This solution should be O(n) time and O(n) space

File: chapter1/javascript/oneAway.js
let oneAway = (str_1, str_2) => {

    // Determine the bigger string

    let len_1 = str_1.length;
    let len_2 = str_2.length;
    if (len_1 > len_2) {
        var big = str_1;
        var small = str_2;
    } else {
        var big = str_2;
        var small = str_1;
    }
    // loop on the longer string
    for (let i = 0; i < len_1; i++) {
        // if these two letters match, continue
        if (big[i] == small[i]) {
            continue
        } else {
            // if the difference is in the first or the last, remove it from both strings
            if (i == 0 || i == (len_1 - 1)) {
                new_str = big.slice(0, i) + big.slice(i + 1, len_1);
                small = small.slice(0, i) + small.slice(i + 1, len_1);
            }
            // if the difference is in the middle, just remove if from the bigger string
            else {
                new_str = big.slice(0, i) + big.slice(i + 1, len_1);
            }
            // now return whether they are equal or not
            return new_str == small;
        }
    }
    // you will only reach this if there is no difference found, so return True
    return true;
}

// TESTS
console.log(oneAway("pale", "ple")) // True
console.log(oneAway("pales", "pale")) // True
console.log(oneAway("pale", "bale")) // True
console.log(oneAway("pale", "bake")) // False
console.log(oneAway("balb", "balp")) // True
console.log(oneAway("balp", "balp")) // True
console.log(oneAway("", "")) // True


// This solution should be O(n) time and O(n) space